### DISCUSSION TITLE
Acesso à dados com Dapper + Domínio Ricos: como trabalhar com entidades?

### DISCUSSION THREAD

**Discussion by rubensmneto** [balta-io/forum#158] [Tira Dúvidas] [answered]
Olá! Estou com uma dúvida interessante.
Eu fiz o curso de Domínios Ricos e agora estou fazendo o de Acesso a Dados com Dapper.
Eu caí num ponto aqui que é o seguinte: no domínios ricos é explicado que eu não posso deixar os setters abertos para alteração externa. Mas agora, no curso de dapper, não foi implementado dessa forma nos exemplos... ou seja, o setter está aberto. Neste caso como faço para alterar, por exemplo, a descrição de um item? Faço um método para isso? Já no construtor do item eu valido se a descrição não está vazia ou muito grande... se eu criar um método eu preciso replicar essa regra? 
Fiquei um pouco confuso com isso... Principalmente quando se faz um update, já que a propriedade não pode ser alterada.

Por exemplo, classe de tipo:

´´´

public class ItemType : Entity<int>
{

    public ItemType (int id, string description) : base(id)
    {

        Description = description;

        if (id <= 0)
        {
            AddNotification(nameof(id), "O ID do tipo não foi informado ou é inválido");
        }

        if (string.IsNullOrEmpty(description))
        {
            AddNotification(nameof(description), "A descrição do tipo não foi informada ou é inválida");
        }

        if (description.Length > 50)
        {
            AddNotification(nameof(description), "O tamanho máximo da descrição do tipo é de 50 caracteres");
        }

    }

    public string Description { get; private set; }

    public override string ToString() => $"{Id} - {Description}";

}

´´´
aí no repositório eu tenho isso

´´´
internal class ItemTypeRepositoryMock : Notifiable, ItemTypeRepository {

         ...

        public Task Update(int id, string description)
        {

            var item = _itemTypes.FirstOrDefault(x => id.Equals(x.Id));

            if (item == null)
            {
                AddNotification(nameof(id), "Tipo de item não encontrado");
                return Task.CompletedTask;
            }

            // aqui está o problema!!
            item.Description = description; // falha porque a propriedade é readonly

            if (!item.IsValid)
            {
                AddNotifications(item.Notifications);
                return Task.CompletedTask;
            }

            return Task.CompletedTask;

        }

        ...

}
´´´

Obrigado pela ajuda pessoal.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @rubensmneto, como vai?

Não é que **não pode deixar os setters abertos**, é que tem abordagens diferentes. Existem abordagens mais voltadas a dados (CRUD) outras mais voltadas a domínio como é o caso do curso que citou.

Não tem certo ou errado, são formas diferentes de fazer a mesma coisa 💜
